scipy/interpolate/regrid_python.py

```diff
@@ -0,0 +1,931 @@
+
+    return np.sum(np.square(R))
+
+def _solve_2d_fitpack(Ax, offs_x, ncx,
+                      Dx, offs_dx,
+                      Ay, offs_y, ncy, Q,
+                      Dy, offs_dy, p,
+                      kx, tx, x_x,
+                      ky, ty, x_y, z):
+    """
+    Solve the 2-D tensor-product spline system using separable banded QR.
+
+    Parameters
+    ----------
+    Ax, Ay : ndarray
+        Banded data matrices for the x and y axes.
+    offs_x, offs_y : ndarray
+        Band offsets for `Ax` and `Ay`.
+    ncx, ncy : int
+        Number of top (data) rows in `Ax` and `Ay`.
+    Dx, Dy : ndarray
+        Banded roughness penalty matrices for x and y.
+    offs_dx, offs_dy : ndarray
+        Band offsets for `Dx` and `Dy`.
+    ncdx, ncdy : int
+        Penalty matrix row counts (from `disc`), informational only.
+    Q : ndarray, shape (mx, my)
+        RHS data grid (copied from `Z`).
+    p : float
+        Smoothing parameter. The penalty term is scaled as **1/p**.
+        Setting `p == -1` signals *p -> inf* (interpolation, omit penalty).
+    kx, ky : int
+        Spline degrees along x and y.
+    tx, ty : ndarray
+        Knot vectors along x and y.
+    x_x, x_y : ndarray
+        Sample coordinates.
+    z : ndarray
+        Original data grid for residual evaluation.
+
+    Returns
+    -------
+    C : ndarray
+        2-D B-spline coefficient grid.
+    fp : float
+        Residual sum of squares between fitted surface and `z`.
+
+    Notes
+    -----
+    This performs two separable QR solves (x then y), each augmented by
+    `(D / p)` when `p != -1`.  Setting `p = -1` skips all penalty rows,
+    yielding an interpolatory surface.  The resulting coefficients and residual
+    follow the same conventions as FITPACK's `fpgrre`.
+    """
+    w_x = np.ones_like(x_x)
+    w_y = np.ones_like(x_y)
+
+    Ax_aug, offset_aug_x, nc_augx = _stack_augmented_fitpack(
+        Ax, offs_x, Dx, offs_dx, ncx, kx, p)
+    Ay_aug, offset_aug_y, nc_augy = _stack_augmented_fitpack(
+        Ay, offs_y, Dy, offs_dy, ncy, ky, p)
+
+    if p != -1:
+        Q = np.vstack([Q, np.zeros((Dx.shape[0], Q.shape[1]), dtype=np.float64)])
+
+    _dierckx.qr_reduce(Ax_aug, offset_aug_x, nc_augx, Q)
+
+    Q_t = np.ascontiguousarray(Q)
+    cT, _, _ = _dierckx.fpback(
+        Ax_aug, nc_augx, x_x,
+        Q_t, tx, kx, w_x,
+        Q_t, False
+    )
+    W = cT.T
+
+    if p != -1:
+        W = np.vstack([W, np.zeros((Dy.shape[0], W.shape[1]), dtype=np.float64)])
+
+    WT = np.ascontiguousarray(W)
+    _dierckx.qr_reduce(Ay_aug, offset_aug_y, nc_augy, WT)
+    W1 = np.ascontiguousarray(WT)
+
+    C, _, fp = _dierckx.fpback(
+        Ay_aug, nc_augy,
+        x_y, W1, ty, ky, w_y,    # y = W1 (dummy)
+        W1,                      # yw = RHS = W1 -> returns C
+        False
+    )
+
+    C = C.T
+    _Ax = BSpline.design_matrix(x_x, tx, kx, extrapolate=False)
+    _Ay = BSpline.design_matrix(x_y, ty, ky, extrapolate=False)
```

Comment:
Is `extrapolate=False` to mimic the lack of extrapolation in FITPACK 2D or is there a different reason? 

A follow-up question is what it takes to allow extrapolation (cf https://github.com/scipy/scipy/issues/3361) : just expose `extrapolate=extrapolate` here or there's more to it?

Reply:
> Is `extrapolate=False` to mimic the lack of extrapolation in FITPACK 2D or is there a different reason?

Fow now yes.



> A follow-up question is what it takes to allow extrapolation (cf [scipy#3361](https://github.com/scipy/scipy/issues/3361)) : just expose `extrapolate=extrapolate` here or there's more to it?

Will explore this once I move this PR from my fork to SciPy repository.